### ISSUE TITLE
[Bug]: Query gets deleted entities if manually compaction

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#17625]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master-20220617-074ec306
- Deployment mode(standalone or cluster): cluster
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus 2.1.0.dev69
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others: 

common:
  retentionDuration: 40
```
```


### Current Behavior

1. Create collection with 1 shard
2. Insert twice and generate two segments
4. Delete 2 entities
5. Insert a new entity
6. sleep `retentionDuration` seconds and call compact 
7.  wait for compact completed, query get the deleted entities

Testcase:
```
    def test_compact_both_delete_merge(self):
        """
        target: test compact both delete and merge
        method: 1.create collection with shard_num=1
                2.insert data into two segments
                3.delete and flush (new insert)
                4.compact
                5.load and search
        expected: Triggre two types compaction
        """
        collection_w = self.init_collection_wrap(cf.gen_unique_str(prefix), shards_num=1)
        ids = []
        for i in range(2):
            df = cf.gen_default_dataframe_data(tmp_nb, start=i * tmp_nb)
            insert_res, _ = collection_w.insert(df)
            assert collection_w.num_entities == (i + 1) * tmp_nb
            ids.extend(insert_res.primary_keys)

        # delete_ids = ids[:tmp_nb]
        delete_ids = [0, tmp_nb // 2]
        expr = f'{ct.default_int64_field_name} in {delete_ids}'
        collection_w.delete(expr)

        collection_w.insert(cf.gen_default_dataframe_data(1, start=2 * tmp_nb))
        assert collection_w.num_entities == 2 * tmp_nb + 1

        sleep(ct.compact_retention_duration + 1)

        collection_w.compact()
        collection_w.wait_for_compaction_completed()
        c_plans = collection_w.get_compaction_plans()[0]

        # assert len(c_plans.plans) == 2
        # todo assert two types compaction plan

        # search
        ids.pop(0)
        ids.pop(-1)
        collection_w.load()
        search_res, _ = collection_w.search(cf.gen_vectors(ct.default_nq, ct.default_dim),
                                            ct.default_float_vec_field_name,
                                            ct.default_search_params, ct.default_limit,
                                            check_items={"nq": ct.default_nq,
                                                         "ids": ids,
                                                         "limit": ct.default_limit})
        collection_w.query(f"{ct.default_int64_field_name} in {delete_ids}", check_task=CheckTasks.check_query_empty)
```
```
<name>: compact_7S4QNeC7
<partitions>: [{"name": "_default", "collection_name": "compact_7S4QNeC7", "description": ""}]
<description>: 
...
[2022-06-17 20:28:55,836 - DEBUG - ci_test]: (api_request)  : [Collection.insert] args: [    int64  float varchar                                       float_vector
0       0    0.0       0  [0.041217323530104315, 0.11628317666822506, 0....
1       1    1.0       1  [0.1271212546595306, 0.015553567340345627, 0.0...
2       2    2.0       2  [0.14161446688445115, 0.10370254049665828, 0......., kwargs: {'timeout': 20} (api_request.py:55)
[2022-06-17 20:28:56,029 - DEBUG - ci_test]: (api_response) : (insert count: 100, delete count: 0, upsert count: 0, timestamp: 433971248748822529)  (api_request.py:27)
[2022-06-17 20:28:59,275 - DEBUG - ci_test]: (api_request)  : [Collection.insert] args: [    int64  float varchar                                       float_vector
0     100  100.0     100  [0.1254693703714313, 0.026068977423383367, 0.0...
1     101  101.0     101  [0.07905920047193112, 0.08943378513043884, 0.1...
2     102  102.0     102  [0.08635146865130318, 0.013475112813227937, 0......, kwargs: {'timeout': 20} (api_request.py:55)
[2022-06-17 20:28:59,420 - DEBUG - ci_test]: (api_response) : (insert count: 100, delete count: 0, upsert count: 0, timestamp: 433971249640112129)  (api_request.py:27)
[2022-06-17 20:29:02,613 - DEBUG - ci_test]: (api_request)  : [Collection.delete] args: ['int64 in [0, 50]', None, 20], kwargs: {} (api_request.py:55)
[2022-06-17 20:29:02,635 - DEBUG - ci_test]: (api_response) : (insert count: 0, delete count: 2, upsert count: 0, timestamp: 433971250492342273)  (api_request.py:27)
[2022-06-17 20:29:02,651 - DEBUG - ci_test]: (api_request)  : [Collection.insert] args: [   int64  float varchar                                       float_vector
0    200  200.0     200  [0.05310856983775027, 0.04262065695172577, 0.1..., None], kwargs: {'timeout': 20} (api_request.py:55)
[2022-06-17 20:29:02,714 - DEBUG - ci_test]: (api_response) : (insert count: 1, delete count: 0, upsert count: 0, timestamp: 433971250505187329)  (api_request.py:27)
[2022-06-17 20:29:46,403 - DEBUG - ci_test]: (api_request)  : [Collection.compact] args: [20], kwargs: {} (api_request.py:55)
[2022-06-17 20:29:46,455 - DEBUG - ci_test]: (api_response) : None  (api_request.py:27)
[2022-06-17 20:29:46,986 - DEBUG - ci_test]: (api_request)  : [Collection.get_compaction_plans] args: [20], kwargs: {} (api_request.py:55)
[2022-06-17 20:29:47,011 - DEBUG - ci_test]: (api_response) : 
Compaction Plans:
 - compaction id: 433971261973987329
 - state: Completed
 - plans: [
Plan:
 - sources: [433971248748822529, 433971249640112129, 433971250505187329]
 - target: 433971261973987331
]
   (api_request.py:27)
[2022-06-17 20:29:47,012 - DEBUG - ci_test]: (api_request)  : [Collection.load] args: [None, 1, 20], kwargs: {} (api_request.py:55)
[2022-06-17 20:29:47,698 - DEBUG - ci_test]: (api_response) : None  (api_request.py:27)
[2022-06-17 20:29:47,700 - DEBUG - ci_test]: (api_request)  : [Collection.search] args: [[[0.03254069517725304, 0.13771638271608624, 0.07469217071437301, 0.08796131231044915, 0.027143802154137548, 0.07948505279576348, 0.11492999337706543, 0.11611379236178723, 0.0017285708107894425, 0.04431899506697454, 0.12781207090330443, 0.02195693449908431, 0.15075795137480033, 0.05568837353279962, ......, kwargs: {} (api_request.py:55)
[2022-06-17 20:29:48,101 - DEBUG - ci_test]: (api_response) : ["['(distance: 0.37805017828941345, id: 121)', '(distance: 0.3811706304550171, id: 99)', '(distance: 0.3858397901058197, id: 97)', '(distance: 0.3868606686592102, id: 151)', '(distance: 0.39088913798332214, id: 55)', '(distance: 0.3956630229949951, id: 44)', '(distance: 0.4060327708721161, id: 36)',......  (api_request.py:27)
[2022-06-17 20:29:48,102 - DEBUG - ci_test]: (api_request)  : [Collection.query] args: ['int64 in [0, 50]', None, None, 20], kwargs: {} (api_request.py:55)
[2022-06-17 20:29:48,265 - DEBUG - ci_test]: (api_response) : [{'int64': 0}, {'int64': 50}]  (api_request.py:27)
FAILED
```

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Milvus Log

_No response_

### Anything else?

_No response_

**Comment by xige-16** [comment]
no delta log when querynode load the compacted segment
```
[2022/06/17 23:13:41.942 +08:00] [INFO] [segment_loader.go:609] ["there are no delta logs saved with segment, skip loading delete record"] [segmentID=433973840138207235]

```

**Comment by xige-16** [comment]
![fHbUwSgvtT](https://user-images.githubusercontent.com/20124155/174333620-566cab17-ff0d-4b9a-ab27-6ea87848a253.jpg)

**Comment by xige-16** [comment]
It seems that delta data may be missed during datanode compaction. @xiaofan-luan

**Comment by xiaofan-luan** [comment]
> It seems that delta data may be missed during datanode compaction. @xiaofan-luan

I will take on look on the compaction part tomorrow~

**Comment by xiaofan-luan** [comment]
The lefted deltalogs should be added at CompleteMergeCompaction in meta.go(Datacoord)
Otherwise the new generated segment should inherit the check point from compacted from segments.
Looking into the delta log part.

**Comment by xiaofan-luan** [comment]
![image](https://user-images.githubusercontent.com/83447078/174419515-c3385e82-abd4-4910-90c9-684a0e425942.png)

**Comment by xiaofan-luan** [comment]
> ![image](https://user-images.githubusercontent.com/83447078/174419515-c3385e82-abd4-4910-90c9-684a0e425942.png)

I thought the dml position should pick the earliest of all segments but not he latest, what do you think? @sunby @xige-16

**Comment by congqixia** [comment]
lgtm, delete msg could be consumed multiple times without affecting the correctness.

**Comment by ThreadDao** [comment]
Fixed `master-20220619-516d9147`
